src/utils/validations/component/validations/with-validation.hoc.js

```diff
@@ -0,0 +1,188 @@
+
+    static propTypes = {
+      children: PropTypes.node,
+      name: PropTypes.string,
+      value: PropTypes.string,
+      validate: PropTypes.func,
+      validations: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      warnings: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      info: PropTypes.oneOfType([
+        PropTypes.func,
+        PropTypes.arrayOf(PropTypes.func)
+      ]),
+      onBlur: PropTypes.func
+    };
+
+    static defaultProps = {
+      validations: [],
+      warnings: [],
+      info: []
+    }
+
+    static contextTypes = {
+      addInput: PropTypes.func,
+      removeInput: PropTypes.func,
+      adjustCount: PropTypes.func
+    }
+
+    state = {
+      hasError: false,
+      hasWarning: false,
+      hasInfo: false
+    };
+
+    currentFailedValidations = { error: [], warning: [], info: [] };
+
+    componentDidMount = () => {
+      if (this.context && this.context.addInput && this.props.validations.length) {
+        this.context.addInput(this.props.name, this.validate);
+      }
+    }
+
+    componentWillUnmount = () => {
+      if (this.context && this.context.removeInput && this.props.validations.length) {
+        this.context.removeInput(this.props.name);
+      }
+    }
+
+    addFailedValidation = (type, message) => {
+      const { adjustCount } = this.context;
+      this.currentFailedValidations[type].push(message);
+      adjustCount(type, 1);
+    }
+
+    removeFailedValidations = (type) => {
+      const array = this.currentFailedValidations[type];
+      const { adjustCount } = this.context;
+      for (let i = array.length - 1; i >= 0; i--) adjustCount(type, -1);
+      this.currentFailedValidations[type] = [];
+    }
+
+    isKnownFailedValidation = (type, error) => {
+      let exists = false;
+      switch (type) {
+        case 'validations':
+          exists = this.currentFailedValidations.error.includes(error.message);
+          break;
+        case 'warnings':
+          exists = this.currentFailedValidations.warning.includes(error.message);
+          break;
+        case 'info':
+          exists = this.currentFailedValidations.info.includes(error.message);
+          break;
+          // no default
+      }
+      return exists;
+    }
+
+    validate = async (types = ['validations', 'warnings', 'info']) => {
+      let result = true;
+      if (types.includes('validations')) result = await this.runValidation('validations');
+      if (result && types.includes('warnings')) result = await this.runValidation('warnings');
+      if (result && types.includes('info')) result = await this.runValidation('info');
+      return result;
+    }
+
+    validationsResult = (type, errorStatus) => {
+      switch (type) {
+        case 'validations':
+          if (errorStatus) {
+            this.addFailedValidation('error', errorStatus.message);
```

Comment:
Good point, I don't think there is a way to avoid race conditions as the way we have validations set up now the Promise rejects on the first validation that fails, and this could be a different one each time if they are all asynchronous.